blockchain/utxocache.go

```diff
@@ -0,0 +1,790 @@
+			return err
+		}
+
+		// Conditionally flush the utxo cache to the database.  Don't force flush
+		// since many blocks may be connected in quick succession when initializing.
+		err = b.utxoCache.MaybeFlush(&n.hash, uint32(n.height), false, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ShutdownUtxoCache flushes the utxo cache to the database on shutdown.  Since
+// the cache is flushed periodically during initial block download and flushed
+// after every block is connected after initial block download is complete,
+// this flush that occurs during shutdown should finish relatively quickly.
+//
+// Note that if an unclean shutdown occurs, the cache will still be initialized
+// properly when restarted as during initialization it will replay blocks to
+// catch up to the tip block if it was not fully flushed before shutting down.
+// However, it is still preferred to flush when shutting down versus always
+// recovering on startup since it is faster.
+//
+// This function should only be called during shutdown.
+func (b *BlockChain) ShutdownUtxoCache() {
+	b.chainLock.RLock()
+	defer b.chainLock.RUnlock()
+
+	tip := b.bestChain.Tip()
+
+	// Force a cache flush and log the flush details.
+	b.utxoCache.MaybeFlush(&tip.hash, uint32(tip.height), true, true)
+}
+
+// FetchUtxoEntry loads and returns the requested unspent transaction output
+// from the point of view of the the main chain tip.
+//
+// NOTE: Requesting an output for which there is no data will NOT return an
+// error.  Instead both the entry and the error will be nil.  This is done to
+// allow pruning of spent transaction outputs.  In practice this means the
+// caller must check if the returned entry is nil before invoking methods on it.
+//
+// This function is safe for concurrent access however the returned entry (if
+// any) is NOT.
+func (b *BlockChain) FetchUtxoEntry(outpoint wire.OutPoint) (*UtxoEntry, error) {
+	b.chainLock.RLock()
+	defer b.chainLock.RUnlock()
+
+	var entry *UtxoEntry
+	err := b.db.View(func(dbTx database.Tx) error {
+		var err error
+		entry, err = b.utxoCache.FetchEntry(dbTx, outpoint)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return entry, nil
+}
+
+// UtxoStats represents unspent output statistics on the current utxo set.
+type UtxoStats struct {
+	Utxos          int64
+	Transactions   int64
+	Size           int64
+	Total          int64
+	SerializedHash chainhash.Hash
+}
+
+// FetchUtxoStats returns statistics on the current utxo set.
+//
+// NOTE: During initial block download the utxo stats will lag behind the best
+// block that is currently synced since the utxo cache is only flushed to the
+// database periodically.  After initial block download the utxo stats will
+// always be in sync with the best block.
+func (b *BlockChain) FetchUtxoStats() (*UtxoStats, error) {
+	var stats *UtxoStats
+	err := b.db.View(func(dbTx database.Tx) error {
+		var err error
+		stats, err = dbFetchUtxoStats(dbTx)
```

Comment:
I think we can save this for another PR, but I think that ultimately the stats should be tracked as part of the cache and flushed along with it so that it is always accurate.  This will become more important when we start looking towards utxo commitments.

Reply:
Yeah, I considered this but ultimately decided to leave it out of this PR.  I'm going to think through the best way to implement it and probably put it in a follow-up PR next.  I initially thought the stats could just be saved along with the `utxoSetState`, but the main complication was efficiently determining the total number of unique transactions (we can't just append for that like we can for the other stats tracking total utxos, size, and amount).